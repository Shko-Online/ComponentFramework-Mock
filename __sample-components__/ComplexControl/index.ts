import type { IInputs, IOutputs } from './generated/ManifestTypes';

export class ComplexControl implements ComponentFramework.StandardControl<IInputs, IOutputs> {
    private boundProperty: boolean;
    private inputProperty: boolean;
    private notifyOutputChanged: () => void;
    private updateCount = 0;
    private container: HTMLDivElement;
    /**
     * Empty constructor.
     */
    constructor() {}

    public async getOutputSchema(context: ComponentFramework.Context<IInputs>): Promise<Record<string, unknown>> {
        return Promise.resolve({
            outputProperty: {
                $schema: 'http://json-schema.org/draft-04/schema#',
                type: 'object',
                properties: {
                    inputProperty: {
                        type: 'boolean',
                    },
                    updateCount: {
                        type: 'number',
                    },
                },
            },
        });
    }

    /**
     * Used to initialize the control instance. Controls can kick off remote server calls and other initialization actions here.
     * Data-set values are not initialized here, use updateView.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to property names defined in the manifest, as well as utility functions.
     * @param notifyOutputChanged A callback method to alert the framework that the control has new outputs ready to be retrieved asynchronously.
     * @param state A piece of data that persists in one session for a single user. Can be set at any point in a controls life cycle by calling 'setControlState' in the Mode interface.
     * @param container If a control is marked control-type='standard', it will receive an empty div element within which it can render its content.
     */
    public init(
        context: ComponentFramework.Context<IInputs>,
        notifyOutputChanged: () => void,
        state: ComponentFramework.Dictionary,
        container: HTMLDivElement,
    ): void {
        this.notifyOutputChanged = notifyOutputChanged;
        // Add control initialization code
        this.container = container;
        this.container.innerHTML = `Updated ${this.updateCount} ${this.updateCount === 1 ? 'time' : 'times'}`;
        this.inputProperty = context.parameters.inputProperty.raw;
        this.boundProperty = context.parameters.boundProperty.raw;
    }

    /**
     * Called when any value in the property bag has changed. This includes field values, data-sets, global values such as container height and width, offline status, control metadata values such as label, visible, etc.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to names defined in the manifest, as well as utility functions
     */
    public updateView(context: ComponentFramework.Context<IInputs>): void {
        // Add control initialization code
        this.updateCount += 1;
        this.container.innerHTML = `Updated ${this.updateCount} ${this.updateCount === 1 ? 'time' : 'times'}`;
        this.boundProperty = context.parameters.boundProperty.raw;
        if (this.inputProperty !== context.parameters.inputProperty.raw) {
            this.inputProperty = context.parameters.inputProperty.raw;
            this.notifyOutputChanged();
        }
    }
    
    /**
     * It is called by the framework prior to a control receiving new data.
     * @returns an object based on nomenclature defined in manifest, expecting object[s] for property marked as “bound” or “output”
     */
    public getOutputs(): IOutputs {
        const updateCount = this.updateCount;
        const boundProperty = this.boundProperty;
        const inputProperty = this.inputProperty;
        const outputProperty = { updateCount, inputProperty };
        return { boundProperty, outputProperty };
    }

    /**
     * Called when the control is to be removed from the DOM tree. Controls should use this call for cleanup.
     * i.e. cancelling any pending remote calls, removing listeners, etc.
     */
    public destroy(): void {
        // Add code to cleanup control if necessary
    }
}
