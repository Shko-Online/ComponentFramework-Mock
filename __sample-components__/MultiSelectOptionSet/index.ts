/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

import { IInputs, IOutputs } from './generated/ManifestTypes';

export class ContainerSize implements ComponentFramework.StandardControl<IInputs, IOutputs> {
    private notifyOutputChanged: () => void;
    private selection: number[];
    private sizeDiv : HTMLDivElement;

    /**
     * Empty constructor.
     */
    constructor() {}

    /**
     * Used to initialize the control instance. Controls can kick off remote server calls and other initialization actions here.
     * Data-set values are not initialized here, use updateView.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to property names defined in the manifest, as well as utility functions.
     * @param notifyOutputChanged A callback method to alert the framework that the control has new outputs ready to be retrieved asynchronously.
     * @param state A piece of data that persists in one session for a single user. Can be set at any point in a controls life cycle by calling 'setControlState' in the Mode interface.
     * @param container If a control is marked control-type='standard', it will receive an empty div element within which it can render its content.
     */
    public init(
        context: ComponentFramework.Context<IInputs>,
        notifyOutputChanged: () => void,
        state: ComponentFramework.Dictionary,
        container: HTMLDivElement,
    ): void {
        // Add control initialization code
    
        const wrapper = document.createElement('form');
        wrapper.style.display = 'flex';
        wrapper.style.flexDirection = 'column';
        this.sizeDiv = document.createElement('div');
        const list = document.createElement('ul');
        list.className = "ShkoOnline.OptionsList";
        context.parameters.selection.attributes?.Options.forEach(option=>{
            const listItem = document.createElement('li');
            const label = document.createElement('label');
            const checkbox = document.createElement('input');
            const tickMark = document.createElement('div');
            tickMark.className = "ShkoOnline.TickMark";
            checkbox.appendChild(tickMark);
            checkbox.type = 'checkbox';
            label.appendChild(checkbox);
            label.append(option.Label);
            listItem.appendChild(label);
            list.appendChild(listItem);
        });
        this.sizeDiv.innerHTML = `${context.mode.allocatedWidth}x${context.mode.allocatedHeight}`;
        wrapper.appendChild(this.sizeDiv);
        container.appendChild(list);
        container.appendChild(wrapper);
        context.mode.trackContainerResize(true);
    }

    /**
     * Called when any value in the property bag has changed. This includes field values, data-sets, global values such as container height and width, offline status, control metadata values such as label, visible, etc.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to names defined in the manifest, as well as utility functions
     */
    public updateView(context: ComponentFramework.Context<IInputs>): void {
        this.sizeDiv.innerHTML = `${context.mode.allocatedWidth}x${context.mode.allocatedHeight}`;
       
        this.selection = context.parameters.selection.raw || [];
    }

    /**
     * It is called by the framework prior to a control receiving new data.
     * @returns an object based on nomenclature defined in manifest, expecting object[s] for property marked as “bound” or “output”
     */
    public getOutputs(): IOutputs {
        const { selection } = this;
        return { selection };
    }

    /**
     * Called when the control is to be removed from the DOM tree. Controls should use this call for cleanup.
     * i.e. cancelling any pending remote calls, removing listeners, etc.
     */
    public destroy(): void {
        // Add code to cleanup control if necessary
    }
}
