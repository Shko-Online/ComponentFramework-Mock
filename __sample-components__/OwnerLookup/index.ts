/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

import { IInputs, IOutputs } from './generated/ManifestTypes';

export class OwnerLookup implements ComponentFramework.StandardControl<IInputs, IOutputs> {
    private container: HTMLDivElement;
    private notifyOutputChanged: () => void;
    private myself: ComponentFramework.LookupValue[];
    private value?: ComponentFramework.LookupValue[];
    /**
     * Empty constructor.
     */
    constructor() {}

    /**
     * Used to initialize the control instance. Controls can kick off remote server calls and other initialization actions here.
     * Data-set values are not initialized here, use updateView.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to property names defined in the manifest, as well as utility functions.
     * @param notifyOutputChanged A callback method to alert the framework that the control has new outputs ready to be retrieved asynchronously.
     * @param state A piece of data that persists in one session for a single user. Can be set at any point in a controls life cycle by calling 'setControlState' in the Mode interface.
     * @param container If a control is marked control-type='standard', it will receive an empty div element within which it can render its content.
     */
    public init(
        context: ComponentFramework.Context<IInputs>,
        notifyOutputChanged: () => void,
        state: ComponentFramework.Dictionary,
        container: HTMLDivElement,
    ): void {
        // Add control initialization code
        this.container = container;
        this.myself = [{ entityType: 'systemuser', id: context.userSettings.userId, name: 'Assigned from PCF' }];
        const name = document.createElement('p');
        this.updateName(context, name);
        container.appendChild(name);

        const button = document.createElement('button');
        button.innerHTML = 'Assign To Myself';
        button.addEventListener('click', this.assignToMyself.bind(this), false);
        container.appendChild(button);
        this.notifyOutputChanged = notifyOutputChanged;
    }

    /**
     * Called when any value in the property bag has changed. This includes field values, data-sets, global values such as container height and width, offline status, control metadata values such as label, visible, etc.
     * @param context The entire property bag available to control via Context Object; It contains values as set up by the customizer mapped to names defined in the manifest, as well as utility functions
     */
    public updateView(context: ComponentFramework.Context<IInputs>): void {
        // Add code to update control view
        const name = this.container.getElementsByTagName('p')[0];
        this.updateName(context, name);
    }

    private updateName(context: ComponentFramework.Context<IInputs>, name: HTMLParagraphElement) {
        const value = context.parameters.value.raw;
        this.value = value;
        name.innerHTML = value && value.length ? `[${value[0].entityType}-${value[0].id}] ${value[0].name}` : '---';
    }

    private assignToMyself() {
        this.value = this.myself;
        this.notifyOutputChanged();
    }

    /**
     * It is called by the framework prior to a control receiving new data.
     * @returns an object based on nomenclature defined in manifest, expecting object[s] for property marked as “bound” or “output”
     */
    public getOutputs(): IOutputs {
        const { value } = this;
        return { value };
    }

    /**
     * Called when the control is to be removed from the DOM tree. Controls should use this call for cleanup.
     * i.e. cancelling any pending remote calls, removing listeners, etc.
     */
    public destroy(): void {
        // Add code to cleanup control if necessary
    }
}
