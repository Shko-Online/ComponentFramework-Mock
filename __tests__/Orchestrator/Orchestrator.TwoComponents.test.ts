/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

import { it, expect, describe, beforeEach } from '@jest/globals';
import {
    ComponentFrameworkMockOrchestrator,
    DateTimePropertyMock,
    EnumPropertyMock,
    LookupPropertyMock,
} from '../../src';
import { OwnerLookup as Control1 } from '../../__sample-components__/OwnerLookup';
import { YearPicker as Control2 } from '../../__sample-components__/YearPicker';
import {
    IInputs as IInputs1,
    IOutputs as IOutputs1,
} from '../../__sample-components__/OwnerLookup/generated/ManifestTypes';

import {
    IInputs as IInputs2,
    IOutputs as IOutputs2,
} from '../../__sample-components__/YearPicker/generated/ManifestTypes';

describe('Orchestrator', () => {
    describe('Owner and YearPicker', () => {
        let orchestrator: ComponentFrameworkMockOrchestrator<[IInputs1, IOutputs1, false, IInputs2, IOutputs2, false]>;
        beforeEach(() => {
            const container1 = document.createElement('div');
            const container2 = document.createElement('div');
            orchestrator = new ComponentFrameworkMockOrchestrator([
                [
                    Control1,
                    {
                        value: LookupPropertyMock,
                    },
                    container1,
                ],
                [
                    Control2,
                    {
                        value: DateTimePropertyMock,
                        enum: EnumPropertyMock<'Yes' | 'No'>,
                    },
                    container2,
                ],
            ]);
            orchestrator.mockGenerators[0].context._SetCanvasItems({
                value: {
                    entityType: 'team',
                    id: 'guid1',
                    name: 'Shko Online',
                },
            });
            orchestrator.mockGenerators[0].context.userSettings.userId = 'MyUserId';
            orchestrator.mockGenerators[0].context.userSettings.userName = 'Betim Beja';
            orchestrator.mockGenerators[1].context._SetCanvasItems({
                value: new Date(2023, 0, 1),
                enum: 'Yes',
            });

            document.body.appendChild(container1);
            document.body.appendChild(container2);
        });

        it('Should render lookup value and date year', () => {
            orchestrator.ExecuteInit();

            orchestrator.mockGenerators[0].ExecuteUpdateView();
            orchestrator.mockGenerators[1].ExecuteUpdateView();

            expect(document.body).toMatchSnapshot();
        });

        it('Should update value on click', () => {
            orchestrator.ExecuteInit();
            orchestrator.mockGenerators[0].ExecuteUpdateView();
            orchestrator.mockGenerators[1].ExecuteUpdateView();

            const button = orchestrator.mockGenerators[0].container.getElementsByTagName('button')[0];
            var evt = document.createEvent('Event');
            evt.initEvent('click', false, true);
            button.dispatchEvent(evt);

            const input = orchestrator.mockGenerators[1].container.getElementsByTagName('input')[0];
            input.value = '2021';
            var evt = document.createEvent('Event');
            evt.initEvent('input', false, true);
            input.dispatchEvent(evt);

            expect(orchestrator.mockGenerators[0].context.updatedProperties).toStrictEqual(['value', 'parameters']);

            expect(orchestrator.mockGenerators[1].context.updatedProperties).toStrictEqual(['value', 'parameters']);
            expect(orchestrator.mockGenerators[1].onOutputChanged.called).toBeTruthy();
            expect(orchestrator.mockGenerators[1].context._parameters.value.raw).toEqual(new Date(2021, 0, 1));

            expect(document.body).toMatchSnapshot();
        });

        afterEach(() => {
            document.body.innerHTML = null;
        });
    });
});
