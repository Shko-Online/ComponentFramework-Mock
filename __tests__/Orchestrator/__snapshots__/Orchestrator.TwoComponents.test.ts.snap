// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orchestrator Owner and YearPicker Should render lookup value and date year 1`] = `
<body>
  <div>
    <p>
      [team-guid1] Shko Online
    </p>
    <button>
      Assign To Myself
    </button>
  </div>
  <div>
    <input
      type="number"
      value="2023"
    />
  </div>
</body>
`;

exports[`Orchestrator Owner and YearPicker Should update value on click 1`] = `
<body>
  <div>
    <p>
      [systemuser-MyUserId] Betim Beja
    </p>
    <button>
      Assign To Myself
    </button>
  </div>
  <div>
    <input
      type="number"
      value="2021"
    />
  </div>
</body>
`;

exports[`Orchestrator Owner and YearPicker Should update values from platform 1`] = `
<body>
  <div>
    <p>
      [team-guid1] Shko Online
    </p>
    <button>
      Assign To Myself
    </button>
  </div>
  <div>
    <input
      type="number"
      value="2023"
    />
  </div>
</body>
`;

exports[`Orchestrator Owner and YearPicker Should update values from platform 2`] = `
<body>
  <div>
    <p>
      ---
    </p>
    <button>
      Assign To Myself
    </button>
  </div>
  <div>
    <input
      type="number"
      value="2022"
    />
  </div>
</body>
`;

exports[`Orchestrator Owner and YearPicker should not have updatedProperties if platform posts same values 1`] = `
<body>
  <div>
    <p>
      [team-guid1] Shko Online
    </p>
    <button>
      Assign To Myself
    </button>
  </div>
  <div>
    <input
      type="number"
      value="2023"
    />
  </div>
</body>
`;

exports[`Orchestrator Owner and YearPicker should not have updatedProperties if platform posts same values 2`] = `
<body>
  <div>
    <p>
      [team-guid1] Shko Online
    </p>
    <button>
      Assign To Myself
    </button>
  </div>
  <div>
    <input
      type="number"
      value="2023"
    />
  </div>
</body>
`;
